edgexfoundry/device-uart
# feat: Upgrade to V3

## PR Checklist
Please check if your PR fulfills the following requirements:

- [Y] I am not introducing a breaking change (if you are, flag in conventional commit message with `BREAKING CHANGE:` describing the break)
- [N] I am not introducing a new dependency (add notes below if you are)
- [N] I have added unit tests for the new feature or bug fix (if not, why?)
- [Y] I have fully tested (add details below) this the new feature or bug fix (if not, why?)
- [N] I have opened a PR for the related docs change (if not, why?)
  <link to docs PR>

## Testing Instructions
1. Run the device-uart service along with edgex-go services. Before that, ensure you're mentioning correct deviceName (i.e., /dev/ttyUSB0) and baudRate (i.e., 115200) in device file and device profile.
2. Read data send from a serial device (I used Docklight serial application) using core-command:
`curl -X GET http://localhost:59882/api/v3/device/name/Uart-Generic-Device/Read_UART | json_pp`
3. Write data to a serial device some data using core-command:
`curl -X PUT -d '{"Write_UART":"57656c636f6d6520746f2074686520456d62656464656420576f726c6421"}'  http://localhost:59882/api/v3/device/name/Uart-Generic-Device/Write_UART | json_pp`
4. Cross verify the device-uart logs (optional).
```
level=INFO ts=2023-06-27T07:50:29.610027287Z app=device-uart source=driver.go:100 msg="Driver.HandleReadCommands(): Received Data =  456d62656464656420506c6174666f726d204c6162"
level=INFO ts=2023-06-27T07:50:06.116906919Z app=device-uart source=driver.go:122 msg="Driver.HandleWriteCommands(): protocols: map[UART:map[baudRate:115200 deviceName:/dev/ttyAMA2]], resource: Write_UART, attribute: map[baud:115200 dev:/dev/ttyAMA2 timeout:1 type:generic], parameters: [DeviceResource: Write_UART, String: 57656c636f6d6520746f2074686520456d62656464656420576f726c6421]"
```

## New Dependency Instructions
+    "github.com/gorilla/mux"
+    "github.com/spf13/cast"

## Additional Info

- I will open a PR for the related docs change after discussion with Lindsey Cheng.
- Attached test logs below:

[Logs_Device-UART-v3_Read-and-Write_Success.txt](https://github.com/edgexfoundry/device-uart/files/11880261/Logs_Device-UART-v3_Read-and-Write_Success.txt)
<img width="960" alt="Docklight_Device-UART-v3_TX-RX-Success" src="https://github.com/edgexfoundry/device-uart/assets/128040806/6f09857a-bf55-42a5-9ab9-1cfe5faa2b28">
